File: binding/src/spatialaudio/libqaudioengine.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qaudioengine.html
pub const qaudioengine = struct {

    /// New constructs a new QAudioEngine object.
    pub fn New() QtC.QAudioEngine {
        return qtc.QAudioEngine_new();
    }


    /// New2 constructs a new QAudioEngine object.
    pub fn New2(parent: ?*anyopaque) QtC.QAudioEngine {
        return qtc.QAudioEngine_new2(@ptrCast(parent));
    }


    /// New3 constructs a new QAudioEngine object.
    pub fn New3(sampleRate: i32) QtC.QAudioEngine {
        return qtc.QAudioEngine_new3(sampleRate);
    }


    /// New4 constructs a new QAudioEngine object.
    pub fn New4(sampleRate: i32, parent: ?*anyopaque) QtC.QAudioEngine {
        return qtc.QAudioEngine_new4(sampleRate, @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioEngine_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudioengine.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetOutputMode(self: ?*anyopaque, mode: i32) void {
        qtc.QAudioEngine_SetOutputMode(@ptrCast(self), @intCast(mode));
    }

    pub fn OutputMode(self: ?*anyopaque, ) i32 {
        return qtc.QAudioEngine_OutputMode(@ptrCast(self));
    }

    pub fn SampleRate(self: ?*anyopaque, ) i32 {
        return qtc.QAudioEngine_SampleRate(@ptrCast(self));
    }

    pub fn SetOutputDevice(self: ?*anyopaque, device: ?*anyopaque) void {
        qtc.QAudioEngine_SetOutputDevice(@ptrCast(self), @ptrCast(device));
    }

    pub fn OutputDevice(self: ?*anyopaque, ) QtC.QAudioDevice {
        return qtc.QAudioEngine_OutputDevice(@ptrCast(self));
    }

    pub fn SetMasterVolume(self: ?*anyopaque, volume: f32) void {
        qtc.QAudioEngine_SetMasterVolume(@ptrCast(self), volume);
    }

    pub fn MasterVolume(self: ?*anyopaque, ) f32 {
        return qtc.QAudioEngine_MasterVolume(@ptrCast(self));
    }

    pub fn SetPaused(self: ?*anyopaque, paused: bool) void {
        qtc.QAudioEngine_SetPaused(@ptrCast(self), paused);
    }

    pub fn Paused(self: ?*anyopaque, ) bool {
        return qtc.QAudioEngine_Paused(@ptrCast(self));
    }

    pub fn SetRoomEffectsEnabled(self: ?*anyopaque, enabled: bool) void {
        qtc.QAudioEngine_SetRoomEffectsEnabled(@ptrCast(self), enabled);
    }

    pub fn RoomEffectsEnabled(self: ?*anyopaque, ) bool {
        return qtc.QAudioEngine_RoomEffectsEnabled(@ptrCast(self));
    }

    pub fn SetDistanceScale(self: ?*anyopaque, scale: f32) void {
        qtc.QAudioEngine_SetDistanceScale(@ptrCast(self), scale);
    }

    pub fn DistanceScale(self: ?*anyopaque, ) f32 {
        return qtc.QAudioEngine_DistanceScale(@ptrCast(self));
    }

    pub fn OutputModeChanged(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_OutputModeChanged(@ptrCast(self));
    }

    pub fn OnOutputModeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QAudioEngine_Connect_OutputModeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn OutputDeviceChanged(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_OutputDeviceChanged(@ptrCast(self));
    }

    pub fn OnOutputDeviceChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QAudioEngine_Connect_OutputDeviceChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MasterVolumeChanged(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_MasterVolumeChanged(@ptrCast(self));
    }

    pub fn OnMasterVolumeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QAudioEngine_Connect_MasterVolumeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn PausedChanged(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_PausedChanged(@ptrCast(self));
    }

    pub fn OnPausedChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QAudioEngine_Connect_PausedChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn DistanceScaleChanged(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_DistanceScaleChanged(@ptrCast(self));
    }

    pub fn OnDistanceScaleChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QAudioEngine_Connect_DistanceScaleChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Start(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_Start(@ptrCast(self));
    }

    pub fn Stop(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_Stop(@ptrCast(self));
    }

    pub fn Pause(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_Pause(@ptrCast(self));
    }

    pub fn Resume(self: ?*anyopaque, ) void {
        qtc.QAudioEngine_Resume(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioEngine_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudioengine.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioEngine_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudioengine.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QAudioEngine_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/audioengine.html#types
pub const enums = struct {
    pub const OutputMode = enum {
        pub const Surround: i32 = 0;
        pub const Stereo: i32 = 1;
        pub const Headphone: i32 = 2;
    };

};

File: binding/src/spatialaudio/libqaudioengine.cpp
#include <QAudioDevice>
#include <QAudioEngine>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qaudioengine.h>
#include "libqaudioengine.h"
#include "libqaudioengine.hxx"

QAudioEngine* QAudioEngine_new() {
	 return new QAudioEngine();
}

QAudioEngine* QAudioEngine_new2(QObject* parent) {
	 return new QAudioEngine(parent);
}

QAudioEngine* QAudioEngine_new3(int sampleRate) {
	 return new QAudioEngine(sampleRate);
}

QAudioEngine* QAudioEngine_new4(int sampleRate, QObject* parent) {
	 return new QAudioEngine(sampleRate, parent);
}

libqt_string QAudioEngine_Tr(const char* s) {
	QString _ret = QAudioEngine::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QAudioEngine_SetOutputMode(QAudioEngine* self, int mode) {
	self->setOutputMode(static_cast<QAudioEngine::OutputMode>(mode));
}

int QAudioEngine_OutputMode(const QAudioEngine* self) {
	return self->outputMode();
}

int QAudioEngine_SampleRate(const QAudioEngine* self) {
	return self->sampleRate();
}

void QAudioEngine_SetOutputDevice(QAudioEngine* self, const QAudioDevice* device) {
	self->setOutputDevice(*device);
}

QAudioDevice* QAudioEngine_OutputDevice(const QAudioEngine* self) {
	return new QAudioDevice(self->outputDevice());
}

void QAudioEngine_SetMasterVolume(QAudioEngine* self, float volume) {
	self->setMasterVolume(volume);
}

float QAudioEngine_MasterVolume(const QAudioEngine* self) {
	return self->masterVolume();
}

void QAudioEngine_SetPaused(QAudioEngine* self, bool paused) {
	self->setPaused(paused);
}

bool QAudioEngine_Paused(const QAudioEngine* self) {
	return self->paused();
}

void QAudioEngine_SetRoomEffectsEnabled(QAudioEngine* self, bool enabled) {
	self->setRoomEffectsEnabled(enabled);
}

bool QAudioEngine_RoomEffectsEnabled(const QAudioEngine* self) {
	return self->roomEffectsEnabled();
}

void QAudioEngine_SetDistanceScale(QAudioEngine* self, float scale) {
	self->setDistanceScale(scale);
}

float QAudioEngine_DistanceScale(const QAudioEngine* self) {
	return self->distanceScale();
}

void QAudioEngine_OutputModeChanged(QAudioEngine* self) {
	self->outputModeChanged();
}

void QAudioEngine_Connect_OutputModeChanged(QAudioEngine* self, intptr_t slot) {
    void (*slotFunc)(QAudioEngine*) = reinterpret_cast<void (*)(QAudioEngine*)>(slot);
    QAudioEngine::connect(self, &QAudioEngine::outputModeChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QAudioEngine_OutputDeviceChanged(QAudioEngine* self) {
	self->outputDeviceChanged();
}

void QAudioEngine_Connect_OutputDeviceChanged(QAudioEngine* self, intptr_t slot) {
    void (*slotFunc)(QAudioEngine*) = reinterpret_cast<void (*)(QAudioEngine*)>(slot);
    QAudioEngine::connect(self, &QAudioEngine::outputDeviceChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QAudioEngine_MasterVolumeChanged(QAudioEngine* self) {
	self->masterVolumeChanged();
}

void QAudioEngine_Connect_MasterVolumeChanged(QAudioEngine* self, intptr_t slot) {
    void (*slotFunc)(QAudioEngine*) = reinterpret_cast<void (*)(QAudioEngine*)>(slot);
    QAudioEngine::connect(self, &QAudioEngine::masterVolumeChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QAudioEngine_PausedChanged(QAudioEngine* self) {
	self->pausedChanged();
}

void QAudioEngine_Connect_PausedChanged(QAudioEngine* self, intptr_t slot) {
    void (*slotFunc)(QAudioEngine*) = reinterpret_cast<void (*)(QAudioEngine*)>(slot);
    QAudioEngine::connect(self, &QAudioEngine::pausedChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QAudioEngine_DistanceScaleChanged(QAudioEngine* self) {
	self->distanceScaleChanged();
}

void QAudioEngine_Connect_DistanceScaleChanged(QAudioEngine* self, intptr_t slot) {
    void (*slotFunc)(QAudioEngine*) = reinterpret_cast<void (*)(QAudioEngine*)>(slot);
    QAudioEngine::connect(self, &QAudioEngine::distanceScaleChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QAudioEngine_Start(QAudioEngine* self) {
	self->start();
}

void QAudioEngine_Stop(QAudioEngine* self) {
	self->stop();
}

void QAudioEngine_Pause(QAudioEngine* self) {
	self->pause();
}

void QAudioEngine_Resume(QAudioEngine* self) {
	self->resume();
}

libqt_string QAudioEngine_Tr2(const char* s, const char* c) {
	QString _ret = QAudioEngine::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QAudioEngine_Tr3(const char* s, const char* c, int n) {
	QString _ret = QAudioEngine::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QAudioEngine_Delete(QAudioEngine* self) {
    delete self;
}

